### PR TITLE
Add authentication service

### PR DESCRIPTION
**Pull request by janakamarasena**
### Proposed changes in this pull request
Introduces the core component required for https://github.com/wso2/product-is/issues/15684

This PR introduces an authentication service. The service atm uses request and response wrappers and internally calls the common auth handler to perform authentication. The service is introduced to provide a JAVA API to perform authentication that is not heavily bound to the HTTP request. While the service currently rely on the HTTP request and response internally the service sends back the authentication decisions using java objects.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceRequestWrapper.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    private Map<String, String[]> parameters = new HashMap<>();
+    private List<AuthenticatorData> authInitiationData = new ArrayList<>();
+
+    public AuthServiceRequestWrapper(HttpServletRequest request, Map<String, String[]> parameters) {
+
+        super(request);
+        this.parameters = parameters;
+        setSessionDataKey(parameters);
+        request.setAttribute(AuthServiceConstants.AUTH_SERVICE_AUTH_INITIATION_DATA, authInitiationData);
+        skipNonceCookieValidation();
```
Comment (janakamarasena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceRequestWrapper.java:50]: Nonce cookie validation skipping is added temporarily and this logic will be improved in another PR. The same protection added by the nonce cookie should be available in some manner

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceRequestWrapper.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        return super.getParameterValues(name);
+    }
+
+    /**
+     * Get the authentication initiation data related to
+     * any authenticators engaged in the request flow.
+     *
+     * @return List of {@link AuthenticatorData} objects.
+     */
+    public List<AuthenticatorData> getAuthInitiationData() {
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceRequestWrapper.java:108]: authInitiationData looks confusing to me, gives the feeling it contains, the data related to authentication flow initiation from the initial request

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceResponseData.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Class containing authentication data that is used to communicate intermediate authentication steps.
+ */
+public class AuthServiceResponseData {
+
+    boolean isAuthenticatorSelectionRequired = false;
+    List<AuthenticatorData> authenticatorOptions = new ArrayList<>();
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceResponseData.java:32]: missing access modifiers

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceResponseWrapper.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    /**
+     * Get the flowId related to the authentication flow.
+     * This is the sessionDataKey of the authentication flow.
+     *
+     * @return String of flowId.
+     * @throws AuthServiceException
+     */
+    public String getFlowId() throws AuthServiceException {
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceResponseWrapper.java:60]: any reason to use this as `flowId` instead of the existing session data key term? We should try to keep the same to avoid any confusions (at least in internal classes, in API level params it would be fine to use a matching name for the api)

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/auth/service/AuthServiceConstants.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+public class AuthServiceConstants {
+
+    /**
+     * Enum for flow status.
+     */
+    public enum FlowStatus {
+
+        SUCCESS_COMPLETED,
+        FAIL_COMPLETED,
+        INCOMPLETE
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/auth/service/AuthServiceConstants.java:33]: don't we have constants for these already?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/auth/service/AuthServiceUtils.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            URI uri = new URI(url);
+            String query = uri.getQuery();
+            String[] pairs = query.split("&");
+            for (String pair : pairs) {
+                int idx = pair.indexOf("=");
+                queryParams.put(URLDecoder.decode(pair.substring(0, idx), UTF_8),
+                        URLDecoder.decode(pair.substring(idx + 1), UTF_8));
+            }
+        } catch (URISyntaxException | UnsupportedEncodingException e) {
+            throw new AuthServiceException("Error while extracting query params from provided url.", e);
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/auth/service/AuthServiceUtils.java:57]: better to mention the URL as part of the message

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java [hunk 7]
```diff
@@ -700,4 +702,13 @@ public static class ActionIDs {
...
         }
     }
+
+    /**
+     * Enum for authenticator param type.
+     */
+    public enum AuthenticatorParamType {
+
+        STRING,
+        INTEGER,
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java:712]: what about boolean, double etc?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionNonceCookieUtil.java [hunk 8]
```diff
@@ -136,4 +136,15 @@ public static boolean isNonceCookieEnabled() {
         return nonceCookieConfig;
     }
 
+    /**
+     * Check if nonce cookie validation should be skipped based on the request.
+     *
+     * @param request Http servlet request.
+     * @return True if nonce cookie validation should be skipped.
+     */
+    public static boolean skipNonceCookieValidation(HttpServletRequest request) {
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionNonceCookieUtil.java:145]: use some name like validateNonceCookie to avoid double negation

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java [hunk 9]
```diff
@@ -121,4 +122,25 @@ default boolean isSatisfyAuthenticatorPrerequisites(HttpServletRequest request,
...
+        return false;
+    }
+
+    /**
+     * Get the authentication initiation data.
+     *
+     * @param context Authentication context.
+     * @return AuthenticatorData containing authentication initiation data.
+     */
+    default AuthenticatorData getAuthInitiationData(AuthenticationContext context) {
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java:142]: use Optional as return type

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        if (isAuthFlowCompleted(request)) {
+            authServiceResponse.setFlowId(getFlowCompletionFlowId(request, response));
+            authServiceResponse.setFlowStatus(AuthServiceConstants.FlowStatus.SUCCESS_COMPLETED);
+            return authServiceResponse;
+        }
+        authServiceResponse.setFlowId(response.getFlowId());
+        authServiceResponse.setFlowStatus(AuthServiceConstants.FlowStatus.INCOMPLETE);
+        setAuthResponseData(request, response, authServiceResponse);
+
+        return authServiceResponse;
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java:98]: what about the fail_completed case?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        List<AuthenticatorData> authenticatorDataList = new ArrayList<>();
+        String[] authenticatorAndIdpsArr = StringUtils.split(authenticatorList,
+                AuthServiceConstants.AUTHENTICATOR_SEPARATOR);
+        for (String authenticatorAndIdps : authenticatorAndIdpsArr) {
+            String[] authenticatorIdpSeperatedArr = StringUtils.split(authenticatorAndIdps,
+                    AuthServiceConstants.AUTHENTICATOR_IDP_SEPARATOR);
+            String name = authenticatorIdpSeperatedArr[0];
+            ApplicationAuthenticator authenticator = FrameworkUtils.getAppAuthenticatorByName(name);
+            if (authenticator == null) {
+                LOG.debug("Authenticator not found for name: " + name);
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java:129]: wrap with isDebugEnabled()

Comment (pulasthi7) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java:129]: and shouldn't this be an error?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+                continue;
+            }
+
+            if (!authenticator.isAPIBasedAuthenticationSupported()) {
+                LOG.debug("Authenticator: " + name + " does not support API based authentication.");
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java:134]: wrap with isDebugEnabled, follow the same for all other similar places

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+                continue;
+            }
+            for (int i = 1; i < authenticatorIdpSeperatedArr.length; i++) {
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java:137]: are we skipping the 0th one in the list? better to add a small comment mentioning what is being done here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceResponseWrapper.java (hunk 4 above)
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceResponseWrapper.java:62]: will change this. This was done thinking of what we want to send outside.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/auth/service/AuthServiceConstants.java (hunk 5 above)
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/auth/service/AuthServiceConstants.java:33]: we have a specific set of constants related to authenticators

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/auth/service/AuthServiceUtils.java (hunk 6 above)
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/auth/service/AuthServiceUtils.java:57]: I was thinking of that. But then thought of avoiding it incase it contained sensitive information that should not be logged.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java (hunk 7 above)
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java:712]: Added these on need basis. Will add those and others when onboarding the relevant authenticators

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionNonceCookieUtil.java (hunk 8 above)
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionNonceCookieUtil.java:145]: wouldn't a method name like validateNonceCookie give the notion that the method performs the validation. thought of renaming to `isNonceCookieValidationSkipped `

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java (hunk 12 above)
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java:134]: Since this debug log doesn't do a costly operation do we need to wrap it in isDebugEnabled()?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 14]
```diff
@@ -673,6 +676,22 @@ protected void doAuthentication(HttpServletRequest request, HttpServletResponse
                     context.getExternalIdP(), authenticator.getName()));
             AuthenticatorFlowStatus status = authenticator.process(request, response, context);
             request.setAttribute(FrameworkConstants.RequestParams.FLOW_STATUS, status);
+            /* If this is an authentication initiation and the authenticator supports API based authentication
+             we need to send the auth initiation data in order to support performing API based authentication.*/
+            if (status == AuthenticatorFlowStatus.INCOMPLETE && authenticator.isAPIBasedAuthenticationSupported()) {
+                AuthenticatorData authInitiationData = authenticator.getAuthInitiationData(context);
+                if (authInitiationData != null) {
```
[the remark this answers is not in the document]
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: fixed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceRequestWrapper.java [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    @Override
+    public Map<String, String[]> getParameterMap() {
+
+        Map<String, String[]> paramMap = new HashMap<>(this.parameters);
+        Map<String, String[]> paramsFromSuper = super.getParameterMap();
+        paramsFromSuper.forEach((key, value) -> {
+            if (!paramMap.containsKey(key)) {
+                paramMap.put(key, value);
+            }
+        });
```
[the remark this answers is not in the document]
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceRequestWrapper.java]: fixed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceRequestWrapper.java [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+
+        return paramMap;
+    }
+
+    @Override
+    public Enumeration<String> getParameterNames() {
+
+        ArrayList<String> paramNames = new ArrayList<>(this.parameters.keySet());
```
[the remark this answers is not in the document]
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceRequestWrapper.java]: fixed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java (hunk 9 above)
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java]: fixed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java (hunk 12 above)
Comment (pulasthi7) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java:134]: yes, this has a string concat, which will contribute to memory growth rate (although by small margin, can add up to a considerable amount if frequently called)

**Comment by janakamarasena** [comment]
Will resolve any merge conflicts when the PR is ready to merge

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+                                           AuthServiceResponse authServiceResponse) throws AuthServiceException {
+
+        authServiceResponse.setSessionDataKey(getFlowCompletionSessionDataKey(request, response));
+        authServiceResponse.setFlowStatus(AuthServiceConstants.FlowStatus.SUCCESS_COMPLETED);
+    }
+
+    private void handleFailedAuthResponse(AuthServiceRequestWrapper request, AuthServiceResponseWrapper response,
+                                          AuthServiceResponse authServiceResponse) throws AuthServiceException {
+
+        // TODO: Improve error handling. Different authenticator seems to pass errors in slightly different ways.
```
Comment (janakamarasena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java:136]: Error handling needs to be further improved as there seems to be several ways errors are communicated. 
This will be improved in a next PR.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+     * @param authRequest The authentication request.
+     * @return The authentication response.
+     * @throws AuthServiceException If an error occurs while handling the authentication request.
+     */
+    public AuthServiceResponse handleAuthentication(AuthServiceRequest authRequest) throws AuthServiceException {
+
+        AuthServiceRequestWrapper wrappedRequest = getWrappedRequest(authRequest.getRequest(),
+                authRequest.getParameters());
+        AuthServiceResponseWrapper wrappedResponse = getWrappedResponse(authRequest.getResponse());
+        CommonAuthenticationHandler commonAuthenticationHandler = new CommonAuthenticationHandler();
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java:68]: this can be class level final variable

Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java]: fixed

**Review by ashensw — APPROVED** [review]
